### PR TITLE
[kimchi][bench] better bench + guideline for flamegraph

### PR DESCRIPTION
**Pull request by mimoo**
* Decreased the number of iterations for benchmarks to 10 (otherwise it takes too much time). 
* Also, separated bench for proving and verifying
* The bench now uses a circuit of 2^16 gates. Should I increase this? I think we might be at 2^17 or 2^18 for Mina. Ideally I would bench different sizes but I don't think our poor-man CI will like that
* The bench uses a circuit made only of multiplication gates. Intuitively I don't think the diversity of gates matter too much so that should be fine (if we're in evaluation form it'll probably won't do much, for operations in coeff form it might).
* Some indications on how to produce a flamegraph

on my mac:

```
Proof creation/proof creation                                                                          
                        time:   [12.283 s 12.825 s 13.418 s]

Proof verification/proof verification                                                                           
                        time:   [248.02 ms 260.22 ms 267.98 ms]
```

on a droplet in digital ocean (8 GB Memory / 4 Intel vCPUs / 160 GB Disk / SFO3 - Debian 11 x64):

```
Proof creation/proof creation (16 gates)
                        time:   [22.907 s 23.156 s 23.394 s]

Proof verification/proof verification
                        time:   [555.30 ms 563.82 ms 569.83 ms]
```

the flamegraph looks like this:

Screen Shot 2022-01-19 at 9 03 11 AM

(verification time is the tiny box)

### CODE REVIEW

**Review by mrmr1993 — CHANGES_REQUESTED** [review]

File: circuits/kimchi/src/gate.rs [hunk 1]
```diff
@@ -280,7 +280,7 @@ impl GateType {
     ///
     /// See circuits/kimchi/src/polynomials/chacha.rs for an explanation of
     /// how these work.
-    #[allow(clippy::type_complexity)]
+    #[allow(clippy::type_complexity)] // TODO: fix this clippy
```
[the remark this answers is not in the document]
Comment (mimoo) [reply circuits/kimchi/src/gate.rs:283]: I was thinking of using a tool to track TODOs, so I still think they should be TODOs (so we don't have to special case the allow(clippy)), but I can create another PRs for that if that's bothering you :D

[the remark this answers is not in the document]
Comment (mimoo) [reply circuits/kimchi/src/gate.rs]: ping

**Comment by mimoo** [comment]
Our verifier circuit is now 2^15, so it might be a good idea to change the bench to 2^15 instead of the 2^16... wdyt?

File: kimchi/src/circuits/gate.rs [hunk 2]
```diff
@@ -279,7 +279,7 @@ impl GateType {
     ///
     /// See circuits/kimchi/src/polynomials/chacha.rs for an explanation of
     /// how these work.
-    #[allow(clippy::type_complexity)]
+    #[allow(clippy::type_complexity)] // TODO: fix this clippy
```
Comment (mrmr1993) [inline kimchi/src/circuits/gate.rs:282]: Delete this TODO comment

File: kimchi/src/circuits/polynomials/lookup.rs [hunk 3]
```diff
@@ -214,7 +214,7 @@ pub fn zk_patch<R: Rng + ?Sized, F: FftField>(
 }
 
 /// Checks that all the lookup constraints are satisfied.
-#[allow(clippy::too_many_arguments)]
+#[allow(clippy::too_many_arguments)] // TODO: fix this clippy
```
Comment (mrmr1993) [inline kimchi/src/circuits/polynomials/lookup.rs:217]: Delete this TODO comment

File: kimchi/src/circuits/polynomials/lookup.rs [hunk 4]
```diff
@@ -491,7 +491,7 @@ pub fn sorted<
 ///
 /// after multiplying all of the values, all of the terms will have cancelled if s is a sorting of f and t, and the final term will be 1
 /// because of the random choice of beta and gamma, there is negligible probability that the terms will cancel if s is not a sorting of f and t
-#[allow(clippy::too_many_arguments)]
+#[allow(clippy::too_many_arguments)] // TODO: fix this clippy
```
Comment (mrmr1993) [inline kimchi/src/circuits/polynomials/lookup.rs:494]: Delete this TODO comment

File: kimchi/src/circuits/polynomials/permutation.rs [hunk 5]
```diff
@@ -20,7 +20,7 @@ use rand::{CryptoRng, RngCore};
 
 impl<F: FftField + SquareRootField> ConstraintSystem<F> {
     /// permutation quotient poly contribution computation
-    #[allow(clippy::type_complexity)]
+    #[allow(clippy::type_complexity)] // TODO: fix this clippy
```
Comment (mrmr1993) [inline kimchi/src/circuits/polynomials/permutation.rs:23]: Delete this TODO comment

File: kimchi/src/circuits/polynomials/varbasemul.rs [hunk 6]
```diff
@@ -29,7 +29,7 @@ fn set<F>(w: &mut [Vec<F>; COLUMNS], row0: usize, var: Variable, x: F) {
     }
 }
 
-#[allow(clippy::too_many_arguments)]
+#[allow(clippy::too_many_arguments)] // TODO: fix this clippy
```
Comment (mrmr1993) [inline kimchi/src/circuits/polynomials/varbasemul.rs:32]: Delete this TODO comment

File: kimchi/src/verifier.rs [hunk 7]
```diff
@@ -46,7 +46,7 @@ where
     /// zeta^n and (zeta * omega)^n
     pub powers_of_eval_points_for_chunks: [Fr<G>; 2],
     /// ?
-    #[allow(clippy::type_complexity)]
+    #[allow(clippy::type_complexity)] // TODO: fix this clippy
```
Comment (mrmr1993) [inline kimchi/src/verifier.rs:49]: Delete this and remaining TODOs in the PR

**Comment by mimoo** [comment]
@mrmr1993 done!

**Review by mrmr1993 — APPROVED** [review]

**Pull request merged**